### PR TITLE
perf: optimize the performance of hyphenate method.

### PR DESCRIPTION
**Pull request by sliwey**
Modify the regular to make the `replace` method call only once to optimize
performance.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [x] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
This is  a performance test:  [https://jsperf.com/vue-hyphenate](https://jsperf.com/vue-hyphenate)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -129,4 +129,4 @@
       "path": "./node_modules/cz-conventional-changelog"
     }
   }
-}
+}
```
Comment (posva) [inline package.json]: can you undo this change?

**Comment by posva** [comment]
Didn't even know about `\b`!
I did some tests and it seems to work nicely 👍

File: package.json (hunk 1 above)
Comment (sliwey) [inline package.json]: @posva  Sorry about that, I have already undo this change :)

**Comment by posva** [comment]
Doing some perf checks and it looks like the old version is faster though 😆

**Comment by sliwey** [comment]
@posva  
I just do a simple test, and this is the [link](https://jsperf.com/vue-hyphenate). It shows the new version is faster, but I only run the test in my environment (chrome 60.0.3112 + windows 10). 
In addition,  regular replacement once should be faster than twice, both the two regular are simple after all (I'm not sure if this is right).
So, how do you do the performance tests?
Sorry for maybe not having made my expression accurately, cause my English is not very good.

**Comment by HerringtonDarkholme** [comment]
Technically this is a breaking change. What counts as a word is listed in http://www.ecma-international.org/ecma-262/5.1/#sec-15.10.2.6

`hyphenateOld('$Naughty')` gets `$-naughty`, the new one gets `$naughty`.

But using dollar sign is not recommended in Vue. I think this change should be good.

**Comment by sliwey** [comment]
@HerringtonDarkholme 
Thanks for your reminder. You're right, the new one can be keep consistent with the old when the parameter belongs to `\w` (it only contains `a-zA-Z0-9_`)

But I'm not sure if the parameter should contain `$` or other unicode characters, so it may need to discuss whether this change is necessary.
